src/tokenserver/db/models.rs

```diff
@@ -1,37 +1,262 @@
-use diesel::mysql::MysqlConnection;
-use diesel::sql_types::Text;
-use diesel::{Connection, RunQueryDsl};
+use actix_web::web::block;
+#[cfg(test)]
+use diesel::sql_types::{Bigint, Integer, Nullable};
+use diesel::{
+    mysql::MysqlConnection,
+    r2d2::{ConnectionManager, PooledConnection},
+    sql_types::Text,
+    RunQueryDsl,
+};
+#[cfg(test)]
+use diesel_logger::LoggingConnection;
+use futures::future::LocalBoxFuture;
+use futures::TryFutureExt;
 
-use super::results::GetTokenserverUser;
+use std::{self, sync::Arc};
+
+use super::{params, results};
 use crate::db::error::{DbError, DbErrorKind};
+use crate::error::ApiError;
+use crate::sync_db_method;
+
+pub type DbFuture<'a, T> = LocalBoxFuture<'a, Result<T, ApiError>>;
+pub type DbResult<T> = std::result::Result<T, DbError>;
+type Conn = PooledConnection<ConnectionManager<MysqlConnection>>;
+
+#[derive(Clone)]
+pub struct TokenserverDb {
+    /// Synchronous Diesel calls are executed in actix_web::web::block to satisfy
+    /// the Db trait's asynchronous interface.
+    ///
+    /// Arc<MysqlDbInner> provides a Clone impl utilized for safely moving to
+    /// the thread pool but does not provide Send as the underlying db
+    /// conn. structs are !Sync (Arc requires both for Send). See the Send impl
+    /// below.
+    pub(super) inner: Arc<DbInner>,
+}
+
+/// Despite the db conn structs being !Sync (see Arc<MysqlDbInner> above) we
+/// don't spawn multiple MysqlDb calls at a time in the thread pool. Calls are
+/// queued to the thread pool via Futures, naturally serialized.
+unsafe impl Send for TokenserverDb {}
+
+pub struct DbInner {
+    #[cfg(not(test))]
+    pub(super) conn: Conn,
+    #[cfg(test)]
+    pub(super) conn: LoggingConnection<Conn>, // display SQL when RUST_LOG="diesel_logger=trace"
+}
+
+impl TokenserverDb {
+    pub fn new(conn: Conn) -> Self {
+        let inner = DbInner {
+            #[cfg(not(test))]
+            conn,
+            #[cfg(test)]
+            conn: LoggingConnection::new(conn),
+        };
+
+        Self {
+            inner: Arc::new(inner),
+        }
+    }
+
+    fn get_user_sync(&self, email: String) -> DbResult<results::GetUser> {
+        let query = r#"
+            SELECT users.uid, users.email, users.client_state, users.generation,
+                users.keys_changed_at, users.created_at, nodes.node
+            FROM users
+            JOIN nodes ON nodes.id = users.nodeid
+            WHERE users.email = ?
+        "#;
+        let mut user_records = diesel::sql_query(query)
+            .bind::<Text, _>(email)
+            .load::<results::GetUser>(&self.inner.conn)?;
+
+        if user_records.is_empty() {
+            return Err(DbErrorKind::TokenserverUserNotFound.into());
+        }
+
+        user_records.sort_by_key(|user_record| (user_record.generation, user_record.created_at));
```

Comment:
We probably want to copy the python version's SQL over directly which will push this sorting into the db layer

Reply:
I think the legacy Tokenserver actually sorts the rows in memory to avoid making the DB indexes too big: https://github.com/mozilla-services/tokenserver/blob/master/tokenserver/assignment/sqlnode/sql.py#L295

Do you think we should change our approach here?

Reply:
Two high-level comments on the legacy tokenserver's approach:

* The cost/benefit tradeoff there is entirely theoretical, I don't think we ever measured the potential bloat of indexes or established that it was too expensive.
* The alleged benefit depends on python's `sort` implementation having almost-linear performance on almost-sorted lists.

FWIW I recommend optimizing for simplicity of implementation, whatever that means in the new system.

Reply:
I'm inclined to say we should just do the sorting in the code. It seems likely that, in the vast majority of cases, a user will only have a handful of old client states, which is what causes "duplicate" user records to be created. Keeping it in Rust seems to me to be less error-prone than writing it in SQL (though I don't think it matters too much). @pjenvey do you have a strong opinion either way?

Reply:
My fault, I was looking solely at the `ORDER BY` clause of the python's get user records call and did not see that it also sorts in code (probably making the former pointless). I'm ok w/ sorting in memory.